### PR TITLE
Discard the following HttpContent for preflight request

### LINKED ISSUE #15148: HttpServerCodec always constructs an "EmptyLastHttpContent" and passes it to the Handler chain when processing OPTIONS requests
HttpServerCodec always constructs an "EmptyLastHttpContent" and passes it to the Handler chain when processing OPTIONS requests, The "EmptyLastHttpContent" propagate through the handler chain.

If a CORS handler is added to the pipeline, it may prematurely return a response to the HTTP client.

However, the CORS handler might still propagate the EmptyLastHttpContent to downstream handlers via fireChannelRead(), causing subsequent handlers to receive only this empty content and lose access to the original request URL.

Can the CORSHandler be configured to directly consume/discard the EmptyLastHttpContent?

simple channel:
```
 ServerBootstrap b = new ServerBootstrap();
                b.group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class)
                    .childHandler(new ChannelInitializer<SocketChannel>() {
                        @Override
                        protected void initChannel(SocketChannel ch) {
                            
                            ch.pipeline().addLast(new HttpServerCodec())

                                .addLast(new CorsHandler(CorsConfigBuilder.forAnyOrigin()
                                    .allowedRequestMethods(HttpMethod.GET, HttpMethod.POST, HttpMethod.PUT, HttpMethod.DELETE, HttpMethod.OPTIONS)
                                    .allowedRequestHeaders("*")
                                    .maxAge(3600L)
                                    .allowNullOrigin()
                                    .build()))

                                .addLast(new SimpleHandler())

```

SimpleHandler: 
```
public SimpleHandler extends ChannelInboundHandlerAdapter {
    private static final Logger log = LoggerFactory.getLogger(SimpleHandler.class);

    private boolean isOption;
 
    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {
         if (msg instanceof HttpRequest req) {
              isOption = req.method() == HttpMethod.OPTIONS;
         }
                                        
          if (msg instanceof LastHttpContent && isOption && msg.toString().equals("EmptyLastHttpContent")) {
                log.info("an EmptyLastHttpContent");
                return;
          }
         ctx.fireChannelRead(msg);
  }
}

```

### PR DESCRIPTION
**Pull request by skyguard1**
Motivation:
HttpServerCodec always constructs an `EmptyLastHttpContent` and passes it to the Handler chain when processing OPTIONS requests, The `EmptyLastHttpContent` propagate through the handler chain.

However, the CORS handler might still propagate the EmptyLastHttpContent to downstream handlers via fireChannelRead(), causing subsequent handlers to receive only this empty content and lose access to the original request URL.
Because `CorsHandler` does not consume this message, it calls `ctx.fireChannelRead(msg)` for the EmptyLastHttpContent.

Downstream handlers then observe:

No HttpRequest

A LastHttpContent with empty content

This often breaks other handlers that rely on receiving the `HttpRequest` first or expect consistent HTTP message.

This PR fixes an issue in `CorsHandler` where, after handling a CORS preflight (OPTIONS) request, the handler still propagates the subsequent `EmptyLastHttpContent` sent by the `HttpServerCodec` to downstream handler.
Modification:

This PR adds option to `CorsHandler` to track when a preflight request has been fully handled.
Track handled preflight and consume the following inbound  HttpContent,LastHttpContent instead of firing it downstream.
Scope: Only affects CORS preflight flow; keeps normal request forwarding unchanged.
Tests: Added cases to verify empty and non-empty LastHttpContent are discarded after preflight and that normal requests continue to forward HttpRequest and content.

It also improves compatibility with application frameworks and routing handlers that expect well-formed HTTP request/response flows.

Result:

Fixes #15148

`CorsHandler` track handled preflight and consume the following HttpContent, LastHttpContent until the next HttpRequest is forwarded  instead of firing it downstream.

### CODE REVIEW

**Comment by normanmaurer** [comment]
hmm... I don't think this is correct. We either need to always pass down a `HttpRequest` + `LastHttpContent` or non of these at all. I don't think there should be any configuration be involved at all.

**Comment by skyguard1** [comment]
> hmm... I don't think this is correct. We either need to always pass down a `HttpRequest` + `LastHttpContent` or non of these at all. I don't think there should be any configuration be involved at all.

However, for preflight requests, the HttpRequest is not being passed down, but the LastHttpContent is. Did I missing something?

**Comment by normanmaurer** [comment]
@skyguard1 I am just saying we either always need to pass down the request + last content or never.. it always needs to be the combination or none at all.

**Comment by skyguard1** [comment]
Preflight (OPTIONS + ACRM) is not a business request; it's merely a control request used internally by the browser to retrieve CORS metadata.

Therefore, the business logic should never:

receive the preflight HttpRequest

receive its HttpContent/LastHttpContent. So, for preflight requests, if you don't pass the HttpRequest, you should also not pass the LastHttpContent. This PR is essentially discarding the LastHttpContent for preflight requests, so I don't know what I'm missing.

**Comment by normanmaurer** [comment]
@skyguard1 What I am saying is that I don't think this should be configurable at all...

**Comment by skyguard1** [comment]
I got it. So you think the flag should be removed, but this fix is ​​valid. Right?

**Comment by skyguard1** [comment]
My idea is to first minimize changes through the configuration, and then remove the flag after the release when users have configured and validated the behavior of discarding LastHttpContent. This satisfies the principle of minimal changes. Of course, we could also directly delete the flag to perform this fix.

**Comment by normanmaurer** [comment]
imho we should just fix it and not have the configuration API. As this is just another API that we need to keep. We could have a system property to fallback to the old behavior but I am not sure this is needed at all.

**Comment by skyguard1** [comment]
Agreed. Then I will delete this configuration.

**Comment by skyguard1** [comment]
Done

File: codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java [hunk 1]
```diff
@@ -87,13 +90,25 @@ public void channelRead(final ChannelHandlerContext ctx, final Object msg) throw
...
         }
+
+        if (discardNextEmptyLast && msg instanceof LastHttpContent) {
+            discardNextEmptyLast = false;
+            LastHttpContent last = (LastHttpContent) msg;
+            if (last == LastHttpContent.EMPTY_LAST_CONTENT) {
+                ReferenceCountUtil.release(last);
+                return;
+            }
+        }
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java:113]: after looking more into the PR I think this is not robust at all and we need to implement this differently. 

So what we really should do is basically keep track if we forward the `HttpRequest` to the next handler and if this is not the case consume / release any following `HttpContent`, `LastHttpContent` until the next `HttpRequest` is forwarded.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 2]
```diff
@@ -512,6 +518,101 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+        // Normal request
+        FullHttpRequest req = new DefaultFullHttpRequest(HTTP_1_1, GET, "/test");
+        req.headers().set(ORIGIN, "http://allowed");
+
+        ch.writeInbound(req);
+
+        HttpContent content = new DefaultHttpContent(Unpooled.copiedBuffer("test message", CharsetUtil.UTF_8));
+        assertFalse(ch.writeInbound(content));
+
+        CaptureHandler cap = ch.pipeline().get(CaptureHandler.class);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: remove the whole `CaptureHandler` stuff and use `ch.readInbound()`. This way you can also verify its the correct type and can also release it

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java (hunk 2 above)
Comment (skyguard1) [reply codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: Done

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 3]
```diff
@@ -512,6 +517,130 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+
+        assertFalse(ch.writeInbound(preflight));
+
+        Object outbound = ch.readOutbound();
+        assertNotNull(outbound); // preflight response
+        ReferenceCountUtil.release(outbound);
+
+        LastHttpContent lastHttpContent = LastHttpContent.EMPTY_LAST_CONTENT;
+        assertFalse(ch.writeInbound(lastHttpContent));
+        ReferenceCountUtil.release(lastHttpContent); // safe release
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: This should be removed... The handler is responsible to release.

```suggestion
```

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 4]
```diff
@@ -512,6 +517,130 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+
+        FullHttpRequest preflight = new DefaultFullHttpRequest(HTTP_1_1, OPTIONS, "/test");
+        preflight.headers().set(ORIGIN, "http://allowed");
+        preflight.headers().set(ACCESS_CONTROL_REQUEST_METHOD, "GET");
+
+        assertFalse(ch.writeInbound(preflight));
+
+        Object outbound = ch.readOutbound();
+        assertNotNull(outbound); // preflight response
+        ReferenceCountUtil.release(outbound);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: This should be removed... The handler is responsible to release.

```suggestion
```

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 5]
```diff
@@ -512,6 +517,130 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+        preflight.headers().set(ACCESS_CONTROL_REQUEST_METHOD, "GET");
+
+        assertFalse(ch.writeInbound(preflight));
+        ReferenceCountUtil.release(ch.readOutbound());
+
+        LastHttpContent first = LastHttpContent.EMPTY_LAST_CONTENT;
+        LastHttpContent second = LastHttpContent.EMPTY_LAST_CONTENT;
+
+        assertFalse(ch.writeInbound(first));
+        ReferenceCountUtil.release(first);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: This should be removed... The handler is responsible to release.


```suggestion
```

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 6]
```diff
@@ -512,6 +517,130 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+
+        assertFalse(ch.writeInbound(second));
+        ReferenceCountUtil.release(second);
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java]: This should be removed... The handler is responsible to release.


```suggestion
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by skyguard1** [comment]
`AutoScalingEventExecutorChooserFactoryTest.testScaleUp`. It seems that the failure of this test case is not related to this PR

**Comment by normanmaurer** [comment]
@yawkat @vietj PTAL

File: codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java [hunk 7]
```diff
@@ -87,13 +90,27 @@ public void channelRead(final ChannelHandlerContext ctx, final Object msg) throw
             config = getForOrigin(origin);
             if (isPreflightRequest(request)) {
                 handlePreflight(ctx, request);
+                // Expect an EmptyLastHttpContent next from HttpServerCodec, mark to discard.
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java]: Please update comment to reflect what the code does now

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsConfigTest.java [hunk 8]
```diff
@@ -141,4 +141,10 @@ public void shortCircuit() {
         final CorsConfig cors = forOrigin("http://localhost:8080").shortCircuit().build();
         assertTrue(cors.isShortCircuit());
     }
+
+    @Test
+    public void allowPrivateNetwork() {
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsConfigTest.java]: nit: This seems to be not related to the PR

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 9]
```diff
@@ -512,6 +517,126 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+        EmbeddedChannel ch = new EmbeddedChannel(new CorsHandler(config));
+
+        FullHttpRequest preflight = new DefaultFullHttpRequest(HTTP_1_1, OPTIONS, "/test");
+        preflight.headers().set(ORIGIN, "http://allowed");
+        preflight.headers().set(ACCESS_CONTROL_REQUEST_METHOD, "GET");
+
+        assertFalse(ch.writeInbound(preflight));
+        ReferenceCountUtil.release(ch.readOutbound());
+
+        LastHttpContent first = LastHttpContent.EMPTY_LAST_CONTENT;
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:555]: nit: You might also want to test with one non-empty one as it should be the same behavior

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Almost there... Please also update the PR description etc to reflect what is actual done now

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsConfigTest.java (hunk 8 above)
Comment (skyguard1) [reply codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsConfigTest.java]: Yes. But I found this allowPrivateNetwork is missing for unit test, maybe remove this?

Comment (normanmaurer) [reply codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsConfigTest.java]: just do it in a separate PR

File: codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java [hunk 10]
```diff
@@ -512,6 +517,141 @@ public void simpleRequestDoNotAllowPrivateNetwork() {
...
+
+        Object first = ch.readInbound();
+        Object second = ch.readInbound();
+
+        assertNotNull(first);
+        assertNotNull(second);
+
+        assertNull(ch.readInbound());
+        assertFalse(ch.finish());
+    }
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:666]: Do we have a test for the case of a preflight request, an empty last http request that isn't forwarded, and then a new http request that _is_ forwarded?

Comment (skyguard1) [reply codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:666]: Sure

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@skyguard1 thanks!

**Comment by skyguard1** [comment]
It is my pleasure
